### PR TITLE
add github webhook filtering

### PR DESCRIPTION
**Pull request by defnot001**
This PR implements the feature that I showcased in [my rust repository](https://github.com/defnot001/testwebhooks) in Typescript. I am not entirly sure how to test this since I am very unfamiliar with serverless infra so if you could maybe test that for me somehow, that would be great.

Thanks!

### CODE REVIEW

File: src/environment.d.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+namespace NodeJS {
+  interface ProcessEnv {
+    APPLICATION_ID: string;
+    DISCORD_TOKEN: string;
+    PUBLIC_KEY: string;
+    GITHUB_SECRET: string;
+    DISCORD_WEBHOOK: string;
+  }
+}
```
Comment (SuperchupuDev) [inline src/environment.d.ts:9]: this file is not needed as cloudflare is not supposed to access env vars from `process.env`, instead update the `Env` type at `src/index.ts`

File: src/gh-webhook/github.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
+import crypto from 'node:crypto';
+import type { Env } from '..';
+
+export async function handleGitHubWebhook(request: Request, env: Env): Promise<Response> {
+  if (request.method !== 'POST') {
+    return new Response('Method not allowed', { status: 405, statusText: 'Method Not Allowed' });
+  }
+
+  const githubSecret = process.env.GITHUB_SECRET;
+  const WebhookURL = process.env.WEBHOOK_URL;
```
Comment (SuperchupuDev) [inline src/gh-webhook/github.ts:10]:
```suggestion
  const githubSecret = env.GITHUB_SECRET;
  const WebhookURL = env.WEBHOOK_URL;
```

File: src/gh-webhook/github.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+  if (!githubSecret || !WebhookURL) {
+    return new Response('Internal server error', { status: 500, statusText: 'Internal Server Error' });
```
Comment (SuperchupuDev) [inline src/gh-webhook/github.ts:13]:
```suggestion
    return new Response('Internal server error', { status: 400, statusText: 'Missing webhook URL or secret' });
```

File: src/gh-webhook/github.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+  if (!sent) {
+    return new Response('Failed to send to Discord', { status: 500, statusText: 'Internal Server Error' });
+  }
+
+  return new Response('Event processed', { status: 200, statusText: 'OK' });
+}
+
+async function isAuthorized(request: Request, githubSecret: string): Promise<boolean> {
+  const header_signature = request.headers.get('x-hub-signature-256');
```
Comment (SuperchupuDev) [inline src/gh-webhook/github.ts:38]: use camelCase for variable naming, i should probably enable the lint rule for that :P

File: .env.example [hunk 5]
```diff
@@ -0,0 +1,5 @@
+APPLICATION_ID=""
+DISCORD_TOKEN=""
+PUBLIC_KEY=""
+GITHUB_SECRET=""
+DISCORD_WEBHOOK="" # make sure to append /github to the end of the webhook URL or else it will not work
```
Comment (SuperchupuDev) [inline .env.example:5]:
```suggestion
APPLICATION_ID=""
DISCORD_TOKEN=""
GITHUB_SECRET=""
PUBLIC_KEY=""
# make sure to append /github to the end of the URL on Discord webhooks
WEBHOOK_URL=""
```
alphabetical order (also that the used env var for the webhook seems to be `WEBHOOK_URL` and not `DISCORD_WEBHOOK`

**Review by SuperchupuDev — CHANGES_REQUESTED** [review]

File: src/gh-webhook/github.ts (hunk 3 above)
Comment (defnot001) [reply src/gh-webhook/github.ts:13]: We can do that but usually I am not a big fan of returning too much information on an endpoint that isn't authenticated yet. This is the bots fault if the variables are incorrect and the 400 ranges usually suggest client error while 500 ranges are for server errors.

File: src/gh-webhook/github.ts (hunk 4 above)
Comment (defnot001) [reply src/gh-webhook/github.ts:38]: Yea I wonder which tool could do something like that :D

File: src/gh-webhook/github.ts (hunk 3 above)
Comment (SuperchupuDev) [reply src/gh-webhook/github.ts:13]: oh wait i misread the line, yeah a 500 for this is good

File: src/gh-webhook/github.ts [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    return new Response('Failed to read request body', { status: 400, statusText: 'Bad Request' });
+  }
+
+  const authorized = await isAuthorized(headers, bodyText, githubSecret);
+
+  if (!authorized) {
+    return new Response('Unauthorized', { status: 401, statusText: 'Unauthorized' });
+  }
+
+  const json = JSON.parse(bodyText).catch(() => null);
```
Comment (SuperchupuDev) [inline src/gh-webhook/github.ts:28]:
```suggestion
  let json: unknown | null;
  try {
    json = JSON.parse(bodyText);
  } catch {
    json = null;
  }
```
`.catch()` only works on [`Promise`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)s

**Review by SuperchupuDev — APPROVED** [review]

**Pull request merged**